src/ints/xms.cpp

```diff
@@ -16,478 +17,843 @@
+	HMA_IN_USE              = 0x91,
+	HMA_NOT_BIG_ENOUGH      = 0x92,
+	HMA_NOT_ALLOCATED       = 0x93,
+	A20_STILL_ENABLED       = 0x94,
+	XMS_OUT_OF_SPACE        = 0xa0,
+	XMS_OUT_OF_HANDLES      = 0xa1,
+	XMS_INVALID_HANDLE      = 0xa2,
+	XMS_INVALID_SRC_HANDLE  = 0xa3,
+	XMS_INVALID_SRC_OFFSET  = 0xa4,
+	XMS_INVALID_DST_HANDLE  = 0xa5,
+	XMS_INVALID_DST_OFFSET  = 0xa6,
+	XMS_INVALID_LENGTH      = 0xa7,
+	XMS_INVALID_OVERLAP     = 0xa8, // TODO: add support for this error
+	XMS_PARITY_ERROR        = 0xa9,
+	XMS_BLOCK_NOT_LOCKED    = 0xaa,
+	XMS_BLOCK_LOCKED        = 0xab,
+	XMS_LOCK_COUNT_OVERFLOW = 0xac,
+	XMS_LOCK_FAILED         = 0xad, // TODO: when should this be reported?
+	UMB_ONLY_SMALLER_BLOCK  = 0xb0,
+	UMB_NO_BLOCKS_AVAILABLE = 0xb1,
+	UMB_INVALID_SEGMENT     = 0xb2, // TODO: when should this be reported?
+};
 
 struct XMS_Block {
-	Bitu	size;
-	MemHandle mem;
-	uint8_t	locked;
-	bool	free;
+	uint32_t size_kb     = 0;
+	MemHandle mem_handle = -1;
+	// locked blocks should not be quietly moved by the XMS driver
+	uint8_t lock_count = 0;
+	bool is_free       = true;
 };
 
 #ifdef _MSC_VER
 #pragma pack (1)
 #endif
-struct XMS_MemMove{
-	uint32_t length;
-	uint16_t src_handle;
+struct XMS_MemMove {
+	uint32_t length = 0;
+
+	uint16_t src_handle = 0;
 	union {
 		RealPt realpt;
 		uint32_t offset;
 	} src;
-	uint16_t dest_handle;
+	uint16_t dst_handle = 0;
 	union {
 		RealPt realpt;
 		uint32_t offset;
-	} dest;
+	} dst;
 
 } GCC_ATTRIBUTE(packed);
 #ifdef _MSC_VER
 #pragma pack ()
 #endif
 
+// ***************************************************************************
+// Variables
+// ***************************************************************************
+
+static struct {
+	bool enable_global = false;
+
+	uint32_t enable_local_count = 0;
+} a20;
+
+constexpr uint32_t A20MaxLocalCount = UINT32_MAX;
+
+static struct {
+	// TODO: HMA support for applications is not yet available in the core
+
+	bool is_available = false;
+
+	bool dos_has_control = true;
+	bool app_has_control = false;
+
+	uint16_t min_alloc_size = 0;
+} hma;
+
+static struct {
+	bool is_available = false;
+} umb;
+
+static struct {
+	bool is_available = false;
+
+	RealPt callback = 0;
 
-Bitu XMS_EnableA20(bool enable) {
-	uint8_t val = IO_Read	(0x92);
-	if (enable) IO_Write(0x92,val | 2);
-	else		IO_Write(0x92,val & ~2);
-	return 0;
+	XMS_Block handles[NumXmsHandles] = {};
+} xms;
+
+// ***************************************************************************
+// Generic helper routines
+// ***************************************************************************
+
+static uint32_t get_num_pages(const uint32_t size_kb)
+{
+	// Number of pages needed for the given memory size in KB
+	return size_kb / 4 + ((size_kb & 3) ? 1 : 0);
+}
+
+static uint32_t get_mem_free_total_kb()
+{
+	return static_cast<uint32_t>(MEM_FreeTotal() * 4);
+}
+
+static uint32_t get_mem_free_largest_kb()
+{
+	return static_cast<uint32_t>(MEM_FreeLargest() * 4);
```

Comment:
Would it make sense to have `MEM_FreeTotal()` (and others) return `uint32_t` too?

Reply:
All this division and multiplication by 4 in multiple functions; can that be extracted into a helper function, or at least extract 4 into a constant with descriptive name?

Reply:
Well, these are the helper functions :) Replaced 4 with a constant where it makes sense.

Reply:
@kcgen We need `uint32_t` actually. Changed.